src/crimson/osd/heartbeat.cc

```diff
@@ -298,12 +298,22 @@ void Heartbeat::heartbeat_check()
     }
 
     if (peer_info.is_unhealthy(now)) {
-      logger().error(" heartbeat_check: no reply from osd.{} "
-		     "since back {} front {} (oldest deadline {})",
-		     osd, peer_info.last_rx_back, peer_info.last_rx_front,
-		     peer_info.ping_history.begin()->second.deadline);
-      failure_queue[osd] = std::min(peer_info.last_rx_back,
-				    peer_info.last_rx_front);
+      auto oldest_deadline = peer_info.ping_history.begin()->second.deadline;
+      if (clock::is_zero(peer_info.last_rx_back) ||
+          clock::is_zero(peer_info.last_rx_front)) {
```

Comment:
how about rewrite this part like:
```c++
auto failed_since = std::min(peer_info.last_rx_back,
                             peer_info.last_rx_front);
if (clock::is_zero(failed_since)) {
  logger().error("heartbeat_check: no reply .. ever ...");
  failed_since = peer_info.first_tx;
} else {
  logger().error("heartbeat_check: no reply .. since ...");
}
failure_queue.emplace(osd, failed_since);
```

Reply:
looks better, changed.